grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+import silver:compiler:definition:type;
+import silver:compiler:definition:type:syntax;
+```
+
+The records extension is meant to codify a common pattern of defining product types that generates efficient code and has nice usability.
+
+## Declaration
+
+A record definition starts with the `record` keyword.
+
+```silver
+terminal RecordKwd_t 'record' lexer classes {KEYWORD};
+```
+
+A record definition looks basically like a nonterminal declaration followed by a list of fields.
+This forwards to a nonterminal, a single production with no children, and an annotation that occurs on the nonterminal for each field.
+
+```silver
+concrete production recordDecl
+this::AGDcl ::= quals::NTDeclQualifiers  'record'  name::IdUpper_t
+  modifiers::NonterminalModifiers  '{'  fields::RecordFields0  '}'  ';'
+{
+  this.unparse =
+    s"${quals.unparse}record ${name.lexeme} ${modifiers.unparse} {${fields.unparse}};";
+
+  local ntDcl::AGDcl =
+    nonterminalDcl(quals, 'nonterminal',
+      nameIdUpper(name, location=this.location),
+      botlNone(location=this.location), modifiers, ';', location=this.location);
+  local ntType::TypeExpr =
+    nominalTypeExpr(qNameTypeId(name, location=this.location),
+      location=this.location);
+
+  -- PARTIAL: head() and tail() are safe because names are nonempty.
+  local nameChars::[String] = explode("", name.lexeme);
+  local prodName::IdLower_t = terminal(IdLower_t,
+    implode("", toLower(head(nameChars)) :: tail(nameChars)),
+    name.location);
+
+  local prodLHS::ProductionLHS = productionLHS(
+    nameIdLower(terminal(IdLower_t, "this"), location=this.location), '::',
+    ntType, location=this.location);
+  local prodSig::ProductionSignature =
+    productionSignatureNoCL(prodLHS, '::=',
+      productionRHSNil(location=this.location), location=this.location);
+  local prodDcl::AGDcl =
+    productionDcl('abstract', 'production',
+      nameIdLower(prodName, location=this.location), prodSig, 
+      productionBody('{', productionStmtsNil(location=this.location), '}',
+        location=this.location),
+      location=this.location);
+
+  forwards to appendAGDcl(ntDcl,
+    appendAGDcl(prodDcl, fields.recordFieldAGDcl, location=this.location),
+    location=this.location);
+}
+```
+
+### `without declaring` Nonterminal Modifier
+
+Note that to support multiple records with the same field names, we don't always declare a new annotation for the field.
+The `without declaring` nonterminal modifier requests that an annotation not be declared.
+It creates an entry in the `withoutDeclaring` synthesized attribute for the annotation.
+
+```silver
+synthesized attribute withoutDeclaring::[String] occurs on NonterminalModifier,
+  NonterminalModifiers, NonterminalModifierList;
+
+terminal Without_kwd   'without'   lexer classes {KEYWORD};
+terminal Declaring_kwd 'declaring' lexer classes {KEYWORD};
+
+concrete production withoutDeclaringOne
+this::NonterminalModifier ::= 'without'  'declaring'  fieldName::IdLower_t
```

Comment:
Why does this reuse NonterminalModifier? I don't think there are any other nonterminal modifiers that we would want to support on records?

Reply:
As discussed, this should probably not be a NonterminalModifier - we don't want to permit this on regular nonterminal declarations.  And I don't see a good reason to declare any other kind of NonterminalModifier on a record nonterminal.

Reply:
Yeah, when I fix the env I'll just drop this section.